rpc/rest/security/rest_security.go

```diff
@@ -0,0 +1,304 @@
+	login = "/login"
+	// URL key for logout, invalidates current token.
+	logout = "/logout"
+)
+
+// AuthenticatorAPI provides methods for handling permissions
+type AuthenticatorAPI interface {
+	// AddPermissionGroup adds new permission group. PG is defined by name and a set of URL keys. User with
+	// permission group enabled has access to that set of keys. PGs with duplicated names are skipped.
+	AddPermissionGroup(group ...*httpsecurity.PermissionGroup)
+
+	// Validate serves as middleware used while registering new HTTP handler. For every request, token
+	// and permission group is validated.
+	Validate(provider http.HandlerFunc) http.HandlerFunc
+}
+
+// Authenticator keeps information about users, permission groups and tokens and processes it
+type authenticator struct {
+	log logging.Logger
+
+	// Router instance automatically registers login/logout REST API handlers if authentication is enabled
+	router    *mux.Router
+	formatter *render.Render
+
+	// User database keeps all known users with permissions and hashed password. Users are loaded from
+	// HTTP config file // todo add other options
+	userDb []*httpsecurity.User
+	// Permission database is a map of permissions and bound URLs
+	groupDb []*httpsecurity.PermissionGroup
+	// Token database keeps information of actual token and its owner.
+	tokenDb map[string]string
+
+	// Token claims
+	expTime int
+}
+
+// NewAuthenticator prepares new instance of authenticator.
+func NewAuthenticator(router *mux.Router, expTime int, users []httpsecurity.User, cost int, log logging.Logger) AuthenticatorAPI {
+	a := &authenticator{
+		router: router,
+		log:    log,
+		formatter: render.New(render.Options{
+			IndentJSON: true,
+		}),
+		userDb:  make([]*httpsecurity.User, 0),
+		tokenDb: make(map[string]string),
+		expTime: expTime,
+	}
+
+	// Add admin-user, enabled by default, always has access to every URL
+	users = append(users, httpsecurity.User{
+		Name:        admin,
+		Password:    "ligato123",
+		Permissions: []string{admin},
+	})
+
+	go func() {
+		// Process users in go routine, since hashing may take some time
+		for _, user := range users {
+			// Password hash cost is currently set via config file. Values allowed by bcrypt are in range 4-31. The min
+			// value will be set if cost is not defined in config file. Keep in mind that high cost values (>10) can
+			// take a lot of time to process.
+			hash, err := bcrypt.GenerateFromPassword([]byte(user.Password), cost)
+			if err != nil {
+				a.log.Errorf("failed to hash password for user %s: %v", user.Name, err)
+				continue
+			}
+
+			a.userDb = append(a.userDb, &httpsecurity.User{
+				Name:        user.Name,
+				Password:    string(hash),
+				Permissions: user.Permissions,
+			})
+			a.log.Debugf("Registered user %s, permissions: %v", user.Name, user.Permissions)
+		}
+	}()
+
+	// Admin-group, available by default and always enabled for all URLs
+	a.AddPermissionGroup(&httpsecurity.PermissionGroup{
```

Comment:
I think there is no need to use exported method for adding default `admin` group, you can just directly initialize the `groupDb` with it when creating `authenticator`.